solr/core/src/java/org/apache/solr/cluster/maintenance/InactiveShardRemover.java

```diff
@@ -0,0 +1,251 @@
+
+  public static final String PLUGIN_NAME = ".inactive-shard-remover";
+
+  static class DeleteActor {
+
+    private final CoreContainer coreContainer;
+
+    DeleteActor(final CoreContainer coreContainer) {
+      this.coreContainer = coreContainer;
+    }
+
+    void delete(final Slice slice, final String asyncId) throws IOException {
+      CollectionAdminRequest.DeleteShard deleteRequest =
+          CollectionAdminRequest.deleteShard(slice.getCollection(), slice.getName());
+      deleteRequest.setAsyncId(asyncId);
+      coreContainer.getZkController().getSolrCloudManager().request(deleteRequest);
+    }
+  }
+
+  private State state = State.STOPPED;
+
+  private final CoreContainer coreContainer;
+
+  private final DeleteActor deleteActor;
+
+  private ScheduledExecutorService executor;
+
+  private long scheduleIntervalSeconds;
+
+  private long ttlSeconds;
+
+  private int maxDeletesPerCycle;
+
+  /** Constructor invoked via Reflection */
+  public InactiveShardRemover(final CoreContainer cc) {
+    this(cc, new DeleteActor(cc));
+  }
+
+  @VisibleForTesting
+  InactiveShardRemover(final CoreContainer cc, final DeleteActor actor) {
+    this.coreContainer = cc;
+    this.deleteActor = actor;
+  }
+
+  @Override
+  public void configure(final InactiveShardRemoverConfig cfg) {
+    this.scheduleIntervalSeconds = cfg.scheduleIntervalSeconds;
+    this.maxDeletesPerCycle = cfg.maxDeletesPerCycle;
+    this.ttlSeconds = cfg.ttlSeconds;
+  }
+
+  @Override
+  public String getName() {
+    return PLUGIN_NAME;
+  }
+
+  @Override
+  public State getState() {
+    return state;
+  }
+
+  @Override
+  public void start() throws Exception {
+    state = State.STARTING;
+    executor = Executors.newScheduledThreadPool(1, new SolrNamedThreadFactory(PLUGIN_NAME));
+    executor.scheduleAtFixedRate(
+        this::deleteInactiveSlices,
+        scheduleIntervalSeconds,
+        scheduleIntervalSeconds,
+        TimeUnit.SECONDS);
+    state = State.RUNNING;
+  }
+
+  @Override
+  public void stop() {
+    if (state == State.RUNNING) {
+      state = State.STOPPING;
+      executor.shutdownNow();
+      try {
+        if (!executor.awaitTermination(10, TimeUnit.SECONDS)) {
+          log.warn(
+              "Executor pool did not terminate within the specified timeout: {} {}",
+              10,
+              TimeUnit.SECONDS);
+        }
+      } catch (InterruptedException e) {
+        log.warn("Failed to shut down the executor pool", e);
+        Thread.currentThread().interrupt();
+      }
+    }
+    state = State.STOPPED;
+  }
+
+  @VisibleForTesting
+  void deleteInactiveSlices() {
+    final ClusterState clusterState = coreContainer.getZkController().getClusterState();
+    Collection<Slice> inactiveSlices = new HashSet<>();
+    clusterState
+        .getCollectionsMap()
+        .forEach((k, v) -> inactiveSlices.addAll(collectInactiveSlices(v)));
```

Comment:
Uniqueness properly will be there without using HashSet but whatever I guess.
BTW I think this stream case is a good use for flatMap.  map DocCollection to its slices (shards) then flatMap them so you have a stream of slices.